packages/mui-system/src/colorManipulator/fast/parse.ts

```diff
@@ -0,0 +1,334 @@
+import MuiError from '@mui/internal-babel-macros/MuiError.macro';
+import { Color, newColor } from './core';
+import * as convert from './convert';
+
+/* eslint-disable no-bitwise */
+
+const HASH = '#'.charCodeAt(0);
+const PERCENT = '%'.charCodeAt(0);
+const G = 'g'.charCodeAt(0);
+const N = 'n'.charCodeAt(0);
+const D = 'd'.charCodeAt(0);
+const E = 'e'.charCodeAt(0);
+
+/**
+ * Approximative CSS colorspace string pattern, e.g. rgb(), color()
+ */
+const PATTERN = (() => {
+  const NAME = '(\\w+)';
+  const SEPARATOR = '[\\s,\\/]';
+  const VALUE = '([^\\s,\\/]+)';
+  const SEPARATOR_THEN_VALUE = `(?:${SEPARATOR}+${VALUE})`;
+
+  return new RegExp(
+    `${NAME}\\(
+      ${SEPARATOR}*${VALUE}
+      ${SEPARATOR_THEN_VALUE}
+      ${SEPARATOR_THEN_VALUE}
+      ${SEPARATOR_THEN_VALUE}?
+      ${SEPARATOR_THEN_VALUE}?
+      ${SEPARATOR}*
+    \\)`.replace(/\s/g, ''),
+  );
+})();
+
+/**
+ * Parse CSS color
+ * Supported formats:
+ * - #hhh, #hhhhhh, #hhhhhhhh
+ * - rgb(), rgba()
+ * - hsl(), hsla()
+ * - color() with one of: 'srgb', 'display-p3', 'a98-rgb', 'prophoto-rgb', 'rec-2020'
+ * @param color CSS color string in one of the supported formats
+ */
+export function parse(color: string): Color {
+  if (color.charCodeAt(0) === HASH) {
+    return parseHex(color);
+  }
+  return parseColor(color);
+}
+
+export function extractHex(hex: string) {
+  let r = 0x00;
+  let g = 0x00;
+  let b = 0x00;
+  let a = 0xff;
+
+  switch (hex.length) {
+    // #59f
+    case 4: {
+      r = (hexValue(hex.charCodeAt(1)) << 4) + hexValue(hex.charCodeAt(1));
+      g = (hexValue(hex.charCodeAt(2)) << 4) + hexValue(hex.charCodeAt(2));
+      b = (hexValue(hex.charCodeAt(3)) << 4) + hexValue(hex.charCodeAt(3));
+      break;
+    }
+    // #5599ff
+    case 7: {
+      r = (hexValue(hex.charCodeAt(1)) << 4) + hexValue(hex.charCodeAt(2));
+      g = (hexValue(hex.charCodeAt(3)) << 4) + hexValue(hex.charCodeAt(4));
+      b = (hexValue(hex.charCodeAt(5)) << 4) + hexValue(hex.charCodeAt(6));
+      break;
+    }
+    // #5599ff88
+    case 9: {
+      r = (hexValue(hex.charCodeAt(1)) << 4) + hexValue(hex.charCodeAt(2));
+      g = (hexValue(hex.charCodeAt(3)) << 4) + hexValue(hex.charCodeAt(4));
+      b = (hexValue(hex.charCodeAt(5)) << 4) + hexValue(hex.charCodeAt(6));
+      a = (hexValue(hex.charCodeAt(7)) << 4) + hexValue(hex.charCodeAt(8));
+      break;
+    }
+    default: {
+      break;
+    }
+  }
+
+  return [r, g, b, a] as [number, number, number, number];
+}
+
+/**
+ * Parse hexadecimal CSS color
+ * @param color Hex color string: #xxx, #xxxxxx, #xxxxxxxx
+ */
+export function parseHex(hex: string): Color {
+  return newColor(...extractHex(hex));
```

Comment:
If `extractHex` allocates a `[r, g, b, a]` array to return it, we're paying an allocation cost on each `parse()` call, and the gains made here are somewhat lost.

It would be better to keep `parseHex` as it was written, and to use the `get(Red|Green|Blue)` function to extract the channels from the color in `colorChannel`:

```js
// in colorChannel()
if (color.charCodeAt(0) === HASH) {
  const c = parseHex(color);
  return `${getRed(c)} ${getGreen(c)} ${getBlue(c)}`;
}
```

Reply:
Got it. I will update the code.